File: src/42_接雨水.cpp
// 42_接雨水
// https://leetcode.cn/problems/trapping-rain-water/description/?envType=study-plan-v2&envId=top-100-liked

#include <vector>
#include <stack>
#include <algorithm>
using namespace std;

// 单调栈写法
class Solution
{
    int trap(vector<int> &height)
    {
        stack<int> stk;
        int ans = 0;
        for (int idx = 0; idx < height.size(); ++idx)
        {
            while (!stk.empty() && height[stk.top()] < height[idx])
            {
                int low = stk.top();
                stk.pop();
                if (!stk.empty())
                {
                    int bounded_height = min(height[idx], height[stk.top()]) - height[low];
                    ans += bounded_height * (idx - stk.top() - 1);
                }
            }
            stk.push(idx);
        }
        return ans;
    }
};

// 双指针
#include <vector>
#include <algorithm> // For std::max and std::min

class Solution
{
public:
    int trap(const std::vector<int> &height)
    {
        std::size_t n = height.size();
        if (n == 0)
        { // Handle empty vector case
            return 0;
        }

        std::vector<int> leftMax(n);
        leftMax[0] = height[0];
        for (std::size_t i = 1; i < n; ++i)
        {
            leftMax[i] = std::max(leftMax[i - 1], height[i]);
        }

        std::vector<int> rightMax(n);
        rightMax[n - 1] = height[n - 1];
        for (std::size_t i = n - 1; i-- > 0;)
        {
            rightMax[i] = std::max(rightMax[i + 1], height[i]);
        }

        int res = 0;
        for (std::size_t i = 0; i < n; ++i)
        {
            res += std::min(rightMax[i], leftMax[i]) - height[i];
        }

        return res;
    }
};

File: src/42_接雨水.py
# 42_接雨水
# https://leetcode.cn/problems/trapping-rain-water/description/?envType=study-plan-v2&envId=top-100-liked

from typing import List

# 单调栈


class Solution:
    def trap(self, height: List[int]) -> int:
        stk = []
        ans = 0
        n = len(height)
        for i in range(n):
            while stk and height[stk[-1]] < height[i]:
                cur_idx = stk.pop()
                if stk:
                    # 注意，这里的宽度计算方式 (i - stk[-1] - 1)
                    ans += (min(height[i], height[stk[-1]]) -
                            height[cur_idx]) * (i - stk[-1] - 1)
            stk.append(i)

        return ans


# s = Solution()
# ans = s.trap([4, 2, 0, 3, 2, 5])
# assert ans == 9


# 双指针写法
# 首先从左往右遍历数组，记录每一个位置左边（包含自己）的最大值，然后再从右往左遍历数组，记录每一个位置右边（包含自己）的最大值。
# 两次遍历完之后我们就知道每一个位置左边和右边的最大值了，左右两边的最大值围成的区域可以看作是一个桶，
# 桶的高度取决于这两个值的最小值，知道桶的高度就可以计算当前位置所能容纳的水了，最后只需要计算所有位置容纳的水量即可。
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        leftMax = height.copy()
        for i in range(1, n):
            leftMax[i] = max(leftMax[i-1], height[i])

        rightMax = height.copy()
        for i in range(n-2, -1, -1):
            rightMax[i] = max(rightMax[i+1], height[i])

        ans = 0
        for i in range(n):
            ans += min(leftMax[i], rightMax[i]) - height[i]

        return ans
# 双指针比单调栈好理解
# 关键是记录每一个位置左边/右边（包含自己）的最大值


s = Solution()
ans = s.trap([4, 2, 0, 3, 2, 5])
assert ans == 9
